---
title: "MNVR report"
author: "C-BIG (CING Bioinformatics Department)"
date: "`r Sys.Date()`"
output: html_document
params:
  datainput: ""
  statsinput: ""
---

```{=html}
<style type="text/css">
.main-container {
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
table.dataTable {
    clear: both;
    margin-top: 6px !important;
    margin-bottom: 6px !important;
    max-width: none !important;
    border-collapse: separate !important;
    border-spacing: 0;
    font-size: smaller;
}
.has-feedback .form-control {
    padding-right: 20.5px;
    font-size: smaller;
}

</style>
```
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning = FALSE, message = FALSE)

```

## Summary Statistics

R Markdown Summary Report. For more details on using R Markdown see <https://github.com/Bioinformatics-Department-C-BIG/MNVR>.

### MNV Statistics
```{r MNVstats, echo=FALSE}
library(DT)
library(tidyverse)
library(gridExtra)
library(karyoploteR)
library(grid)
dft<-params$datainput

nsamples<-length(unique(dft$sample))
dfs<-dft[,c(1,2,13,14,15)]%>%group_by(sample,consequence) %>% summarise_each(n_distinct)

dfst<-dft[,c(1,2,13,14,15)] %>% summarise_all(n_distinct)



dfs<-as.data.frame(rbind.data.frame(dfst,dfs))
dfs[1,c(1,ncol(dfs))]<-c('All (n)','All (n)')

dfst<-dft[,c(1,2,13,14,15)]%>%group_by(sample) %>% summarise_each(n_distinct)%>% summarise_all(mean)


dfs<-as.data.frame(rbind.data.frame(dfst,dfs))

dfs[1,c(1,ncol(dfs))]<-c('All (mean)','All (mean)')
# colnames(dfs)<-c('Sample','Chromosomes (n)','Genes (n)','Transcripts (n)','Consequence')


dt<-DT::datatable(dfs[,-1],rownames = dfs$sample,autoHideNavigation = T,extensions = 'Buttons',
                  options = list(pageLength=10,autoWidth = TRUE,
                                 dom = 'Bfrtip',
                                 buttons = list('pageLength', 'colvis','excel','csv',list(
                                                            extend = 'pdf',
                                                            pageSize = 'A2',
                                                            orientation = 'landscape',
                                                            filename = 'tt'
                                                    ))
                                 )
                  )%>%DT::formatStyle(columns = colnames(dfs)[-1], fontSize = '75%')
dt

```
### Phasing Statistics

```{r Phasingstats, echo=FALSE}

Sdf<-params$statsinput
# 
# nsamples<-length(unique(dft$sample))
# dfs<-dft[,c(1,2,13,14,15)]%>%group_by(sample,consequence) %>% summarise_each(n_distinct)
# 
# dfst<-dft[,c(1,2,13,14,15)] %>% summarise_all(n_distinct)
# 
# 
# 
# dfs<-as.data.frame(rbind.data.frame(dfst,dfs))
# dfs[1,c(1,ncol(dfs))]<-c('All (n)','All (n)')
# 
# dfst<-dft[,c(1,2,13,14,15)]%>%group_by(sample) %>% summarise_each(n_distinct)%>% summarise_all(mean)
# 
# 
# dfs<-as.data.frame(rbind.data.frame(dfst,dfs))
# 
# dfs[1,c(1,ncol(dfs))]<-c('All (mean)','All (mean)')
# # colnames(dfs)<-c('Sample','Chromosomes (n)','Genes (n)','Transcripts (n)','Consequence')


dt<-DT::datatable(Sdf,autoHideNavigation = T,extensions = 'Buttons',
                  options = list(pageLength=10,autoWidth = TRUE,
                                 dom = 'Bfrtip',
                                 buttons = list('pageLength', 'colvis','excel','csv',list(
                                                            extend = 'pdf',
                                                            pageSize = 'A2',
                                                            orientation = 'landscape',
                                                            filename = 'tt'
                                                    ))
                                 )
                  )%>%DT::formatStyle(columns = colnames(Sdf)[-1], fontSize = '75%')
dt

```




## Plots

Summary plots for batch:

```{r plots, echo=FALSE,fig.width=18,fig.height=12}

p1<-ggplot(dft,aes(y=consequence,fill=consequence))+geom_bar(stat = 'count',show.legend = F)+theme_bw()+theme_bw()+scale_fill_brewer(palette='Set2',direction = -1)

p2<-ggplot(dft,aes(y=POS-POS2,fill=consequence))+geom_bar(stat = 'count',show.legend = T)+theme_bw()+ylab('Distance between SNPs (bp)')+theme_bw()+scale_fill_brewer(palette='Set2',direction = -1)



names<-names(table(dft$gene)[order(table(dft$gene),decreasing = T)][1:20])
dfg<-dft[which(dft$gene%in%names),]

p3<-ggplot(dfg,aes(y=gene,fill=consequence))+geom_bar(stat = 'count',show.legend = T,alpha = 0.8)+theme_bw()+scale_fill_brewer(palette='Set2',direction = -1)




dfs<-dft
dfs$POS<-ifelse(dft$POS<=dft$POS2,dft$POS,dft$POS2)
dfs$POS2<-ifelse(dft$POS<=dft$POS2,dft$POS2,dft$POS)
dfs$CHR<-paste0('chr',dfs$CHR)


gr<-makeGRangesFromDataFrame(dfs,keep.extra.columns = F,ignore.strand = T,seqnames.field = 'CHR',seqinfo = Seqinfo(unique(dfs$CHR)),start.field = 'POS',end.field = 'POS2')

  ccod<-list('missense'='orange','synonymous'='royalblue','stop_gained'='orangered')
 p4<- as.ggplot(expression(plotKaryotype()%>% karyoploteR::kpPlotDensity(data=gr,data.panel = 2,col="#CCCCFF") %>%
    karyoploteR::kpPoints(chr = paste0('chr',dft$CHR),x=dft$POS,y=0.2, col=as.vector(unlist(ccod[dft$consequence])))))
 
 
 grid.arrange(grobs=list(p1,p2,p3,p4),layout_matrix=matrix(c(1,3,2,3,4,4,4,4),ncol=4))

```
## Explicit Results Table

Explicit MNV entries:
```{r explicitdt, echo=FALSE}
 
dt<-DT::datatable(dft[c(1:50),],rownames = F,autoHideNavigation = T,extensions = 'Buttons',
                  options = list(pageLength=20,autoWidth = TRUE,
                                 dom = 'Bfrtip',
                                 buttons = list('pageLength', 'colvis','excel','csv',list(
                                                            extend = 'pdf',
                                                            pageSize = 'A2',
                                                            orientation = 'landscape',
                                                            filename = 'tt'
                                                    ))
                                 ),
                  filter = 'top')%>%DT::formatStyle(columns = colnames(dft), fontSize = '75%')
dt


```
